src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp

```diff
@@ -146,46 +146,45 @@ bool AsynchronousReadIndirectBufferFromRemoteFS::nextImpl()
         return false;
 
     size_t size = 0;
-
     if (prefetch_future.valid())
     {
         ProfileEvents::increment(ProfileEvents::RemoteFSPrefetchedReads);
 
-        CurrentMetrics::Increment metric_increment{CurrentMetrics::AsynchronousReadWait};
-        Stopwatch watch;
+        size_t offset = 0;
         {
+            Stopwatch watch;
+            CurrentMetrics::Increment metric_increment{CurrentMetrics::AsynchronousReadWait};
             auto result = prefetch_future.get();
             size = result.size;
-            auto offset = result.offset;
-            assert(offset < size);
-
-            if (size)
-            {
-                memory.swap(prefetch_buffer);
-                size -= offset;
-                set(memory.data() + offset, size);
-                working_buffer.resize(size);
-                file_offset_of_buffer_end += size;
-            }
+            offset = result.offset;
+            assert(offset < size || size == 0);
+            ProfileEvents::increment(ProfileEvents::AsynchronousReadWaitMicroseconds, watch.elapsedMicroseconds());
         }
 
-        watch.stop();
-        ProfileEvents::increment(ProfileEvents::AsynchronousReadWaitMicroseconds, watch.elapsedMicroseconds());
+        if (size)
+        {
+            prefetch_buffer.swap(memory);
+            file_offset_of_buffer_end += size;
+            BufferBase::set(memory.data(), size, offset);
+            /// Adjust the working buffer so that it starts at `offset`.
+            working_buffer = Buffer(working_buffer.begin() + offset, working_buffer.end());
```

Comment:
I think previous version with `set(memory.data() + offset, size);` was better than these two lines.

Reply:
The reason to refactor like this is because we can refer to `internal_buffer.size()` in some place, such as https://github.com/ClickHouse/ClickHouse/blob/master/src/IO/AsynchronousReadBufferFromFileDescriptor.cpp#L66

It's the only advantage I can image based on the current BufferBase implementation: separate working_buffer and internal_buffer.

internal_buffer always holds the entire range of the available memory.

working_buffer can shrink into a smaller span of memory.

So a lazy seek operation with bytes_to_ignore can be represented as following:

```
internal_buffer: |__________________|

working_buffer:  |xxxxx|____________|
                 ^     ^
              bytes_to_ignore         
```

And yes, the interface can be improved. We can have dedicated ignore method for this operation.